libs/etcd_broker/src/lib.rs

```diff
@@ -0,0 +1,288 @@
+                    "Timeline updates subscription exited with errors, kind: {:?}",
+                    self.kind
+                )
+            })?;
+        Ok(())
+    }
+}
+
+/// The subscription kind to the timeline updates from safekeeper.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum SkTimelineSubscriptionKind {
+    /// Get every timeline update.
+    All,
+    /// Get certain tenant timelines' updates.
+    Tenant(ZTenantId),
+    /// Get certain timeine updates.
+    Timeline(ZTenantTimelineId),
+}
+
+impl SkTimelineSubscriptionKind {
+    fn watch_regex(&self) -> Regex {
+        match self {
+            Self::All => Regex::new(&format!(
+                r"^{ZENITH_BROKER_PREFIX}/([[:xdigit:]]+)/([[:xdigit:]]+)/safekeeper/([[:digit:]])$"
+            ))
+            .expect("wrong regex for 'everything' subscription"),
+            Self::Tenant(tenant_id) => Regex::new(&format!(
+                r"^{ZENITH_BROKER_PREFIX}/{tenant_id}/([[:xdigit:]]+)/safekeeper/([[:digit:]])$",
+            ))
+            .expect("wrong regex for 'tenant' subscription"),
+            Self::Timeline(ZTenantTimelineId {
+                tenant_id,
+                timeline_id,
+            }) => Regex::new(&format!(
+                r"^{ZENITH_BROKER_PREFIX}/{tenant_id}/{timeline_id}/safekeeper/([[:digit:]])$",
+            ))
+            .expect("wrong regex for 'timeline' subscription"),
+        }
+    }
+
+    /// Etcd key to use for watching a certain timeline updates from safekeepers.
+    pub fn watch_key(&self) -> String {
+        match self {
+            Self::All => ZENITH_BROKER_PREFIX.to_string(),
+            Self::Tenant(tenant_id) => {
+                format!("{ZENITH_BROKER_PREFIX}/{tenant_id}/safekeeper")
+            }
+            Self::Timeline(ZTenantTimelineId {
+                tenant_id,
+                timeline_id,
+            }) => format!("{ZENITH_BROKER_PREFIX}/{tenant_id}/{timeline_id}/safekeeper"),
+        }
+    }
+}
+
+/// Creates a background task to poll etcd for timeline updates from safekeepers.
+/// Stops and returns `Err` on any error during etcd communication.
+/// Watches the key changes until either the watcher is cancelled via etcd or the subscription cancellation handle,
+/// exiting normally in such cases.
+pub async fn subscribe_to_timeline_updates(
+    client: &mut Client,
+    subscription_kind: SkTimelineSubscriptionKind,
+) -> anyhow::Result<SkTimelineSubscription> {
+    info!("Subscribing to timeline updates, subscription kind: {subscription_kind:?}");
+
+    let (watcher, mut stream) = client
+        .watch(
+            subscription_kind.watch_key(),
+            Some(WatchOptions::new().with_prefix()),
+        )
+        .await
+        .with_context(|| {
+            format!("Failed to init the watch for subscription {subscription_kind:?}")
+        })?;
+
+    let (timeline_updates_sender, safekeeper_timeline_updates) = mpsc::unbounded_channel();
```

Comment:
I don't understand the purpose of this channel. Let me try to explain how I thought this would work:

The WAL receiver wants to wake up, when the state in etcd changes. To do that, we create an etcd watch. The watch is a future that returns when something changed for the timeline we're interested in.

In the WAL receiver loop, we simultenously wait for the etcd watch, and for the I/O on the libpq connection.

What do we need the channel for?

Reply:
I've pushed a few more docs that aim to describe the current impl state, but indeed I find the current approach a bit hard too, so I'd love to refactor it more before adding any proper docs.

The

> we simultenously wait for the etcd watch

is the one that I'm not sure how to implement easier: I'm aware of the following tokio future manipulation primitives:
* `.await` and nothing after this in the block can be executed
* `tokio::select!` that runs multiple futures and processes the one that's run first, "cancelling" the rest
There's a ["cancellation safety"](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety) notion that limits what could safely be placed in such `select!` and I'm not sure that wal streaming with all the buffers is cancellation safe, since the tasks get abruptly shut down
* `tokio::join!`/`FuturesOrdered`/`FuturesUnordered` for concurrently running _all_ the tasks to the end.
* `tokio::spawn`/`tokio::task::spawn_blocking` for "run it on a separate thread" analogue, with similar Send + Sync + 'static requirements

So there are multiple async sources of data, each with its own cadence and potentially blocking behaviour:
* `tenant_mgr` populates its timeline map and notifies the main WAL receiver thread about this. Same for the removal
Handed in a top level wal receiver thread, since has to spawn timeline-wise broker loops
* `etcd` that's very swift to produce messages, minimum one per second, now handled in the broker loop
* the actual WAL stream from safekeeper, that could take a while to open, will hopefully stay open for a long time, if the safekeeper we're connected to is healthy

So it looks like if we place the WAL streaming task to `select!` with etcd and the rest, it will get cancelled too quickly due to etcd message storm.
If we place it to `join!`, then nothing gets updated before the entire wal stream loop finishes.
So I've moved away the WAL loop into a separate tokio task via `spawn!`, it's all the existing WAL streaming code basically, it was almost async already.
We need to react on the task finish, to restart a connection immediately hence one of the channels + the other one to cancel the entire WAL receiving (if we abort the connection).

etcd subscriptio + connection management lives one level up, starts per timeline and also needs some channel to signal that it has to stop: has to be different from the WAL stop signal, since broker and safekeeper connections are managed separately.

then, the top level watched of the channel that gives us new tenant tasks (add/remove), so that we could start or stop WAL streaming for those.


Unless I'm missing some other good way to concurrently poll tasks (without manual state machine writing), I think that the "nested control loops" structure will stay, but will get wrapped into more readable structs.
What I want to remove are the channels, and also investirage, if `tokio::task::JoinHandle` cancellation is a fine thing to do: then indeed we can reduce the number of channels by `select!` on join handles directly.

Reply:
My understanding is the same, i.e. that running communication with broker in separate task is the most sensible way.

> tokio::task::JoinHandle cancellation is a fine thing to do

From the common sense I suppose so, but don't see a confirmation in the docs, so better check.

Reply:
Had worked on that more, there's a way to await on the handle as `(&mut handle).await` and it seems to behave ok and not to cancel the task after the future drop.

But it's not really needed with the current model, when I have to get periodic events from the WAL receiver task to ensure we keep the connection up-to-date.